Sources/DipError.swift

```diff
@@ -89,6 +98,87 @@ public enum DipError: Error, CustomStringConvertible {
       return "Resolved instance \(resolved ?? "nil") does not implement expected type \(key.type)."
     }
   }
+}
+
+extension DipError {
   
+  /// Informs you if the resolve failed because recursive limit was reached. Warning: O(n) for depth. May be slow
+  public var isRecusiveError: Bool {
+    switch self {
+      case let .autoWiringFailed(type: _, underlyingError: underlyingError):
+        if let dipError = underlyingError as? DipError {
+          return dipError.isRecusiveError
+        }
+        return false
+      case .recursionDepthReached:
+        return true
+      default:
+        return false
+    }
+  }
+  
+  public struct RecuriveErrorReport: CustomStringConvertible {
+    //The enture resolve stack that resulted in the recursive depth error
+    public let resolveStack: [Any.Type]
+    
+    //The sub stack that was identified as a cycle, if any.
+    public let cycleSubStack: [Any.Type]
+    
+    public var description: String {
+      if(cycleSubStack.count > 0) {
+        return "Recursive Depth exceeded: Cycle Found: \(cycleSubStack)"
+      }
+      return "Recursive Depth exceeded. No cycle found (inconclusive). Full stack: \(resolveStack)"
+    }
+  }
+  
+  public func analyzeRecursiveError() -> RecuriveErrorReport? {
+    if !isRecusiveError {
+      return nil
+    }
+    
+    var resolveStack = [Any.Type]()
+    
+    func buildStack(error: DipError) {
+      switch error {
+        case let .autoWiringFailed(type: type, underlyingError: underlyingError):
+          resolveStack.append(type)
+          if let dipError = underlyingError as? DipError {
+              buildStack(error: dipError)
+          }
+          return
+        case .recursionDepthReached:
+          return
+        default:
+          return
+      }
+    }
+    
+    buildStack(error: self)
+    
+    //Search for cycle subStack
+    var cycleSubStack = [Any.Type]()
+    var hitCount = [String:Int]()
+    for (index, element) in resolveStack.enumerated() {
+      let entryText = String(reflecting: element)
+      let hit = (hitCount[entryText] ?? 0) + 1
+      if(hit == 2) {
+        let _last = resolveStack[0..<index].lastIndex { (type) -> Bool in
+                   return String(reflecting: type) == entryText
+               }
+        guard let last = _last else {
+          break
+        }
```

Comment:
also minor, but this could be combined for readability:
```swift
guard let last = resolveStack[0..<index].lastIndex(where: { (type) -> Bool in
        return String(reflecting: type) == entryText
    }) else {
        break
    }
```